File: sdk/python/maestro_sdk/csiks.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any, Dict, List, Optional

import httpx


class CsiksError(Exception):
    """Raised when the CSIKS service returns a non-success response."""

    def __init__(self, status_code: int, message: str, details: Optional[Any] = None) -> None:
        super().__init__(message)
        self.status_code = status_code
        self.details = details


@dataclass
class IssueRequest:
    tenant: str
    action: str
    resource: str
    ttl_seconds: Optional[int] = None
    issued_by: Optional[str] = None
    metadata: Optional[Any] = None


@dataclass
class IssueResponse:
    idempotency_key: str
    expires_at: str
    status: str


@dataclass
class VerifyRequest:
    tenant: str
    action: str
    resource: str
    idempotency_key: str
    fingerprint: Optional[str] = None
    participant: Optional[str] = None
    dedupe_token: Optional[str] = None


@dataclass
class VerifyResponse:
    status: str
    dedupe_count: int
    expires_at: str
    last_seen: str
    conflict_reason: Optional[str]
    replay_hint: Optional[str]
    result: Optional[Any]
    journal: List['JournalEntry']


@dataclass
class CompleteRequest:
    tenant: str
    action: str
    resource: str
    idempotency_key: str
    participant: Optional[str] = None
    success: bool = True
    result: Optional[Any] = None
    note: Optional[str] = None


@dataclass
class CompleteResponse:
    status: str
    result: Optional[Any]
    journal: List['JournalEntry']


@dataclass
class RecordView:
    status: str
    tenant: str
    action: str
    resource: str
    fingerprint: Optional[str]
    dedupe_count: int
    created_at: str
    last_seen: str
    expires_at: str
    conflict_reason: Optional[str]
    result: Optional[Any]
    journal: List['JournalEntry']


@dataclass
class JournalEntry:
    timestamp: str
    participant: Optional[str]
    event: str
    note: Optional[str]
    fingerprint: Optional[str]
    dedupe_count: int
    details: Optional[Any]


class CsiksClient:
    """Minimal async client for the Cross-System Idempotency Key Service."""

    def __init__(
        self,
        base_url: str,
        *,
        headers: Optional[Dict[str, str]] = None,
        timeout: float = 10.0,
        client: Optional[httpx.AsyncClient] = None,
    ) -> None:
        self._base_url = base_url.rstrip('/')
        self._owns_client = client is None
        self._client = client or httpx.AsyncClient(
            base_url=self._base_url,
            timeout=timeout,
            headers=headers,
        )

    async def aclose(self) -> None:
        if self._owns_client:
            await self._client.aclose()

    async def issue_key(self, request: IssueRequest) -> IssueResponse:
        response = await self._request('POST', '/keys/issue', json=_drop_none(asdict(request)))
        return _parse_issue_response(response.json())

    async def verify_key(self, request: VerifyRequest) -> VerifyResponse:
        response = await self._request('POST', '/keys/verify', json=_drop_none(asdict(request)))
        return _parse_verify_response(response.json())

    async def complete_key(self, request: CompleteRequest) -> CompleteResponse:
        response = await self._request('POST', '/keys/complete', json=_drop_none(asdict(request)))
        return _parse_complete_response(response.json())

    async def get_record(self, key: str) -> RecordView:
        response = await self._request('GET', f'/keys/{key}')
        return _parse_record_view(response.json())

    async def _request(self, method: str, path: str, **kwargs: Any) -> httpx.Response:
        response = await self._client.request(method, path, **kwargs)
        if response.is_error:
            try:
                details = response.json()
                message = details.get('error', response.text)
            except ValueError:
                details = None
                message = response.text
            raise CsiksError(response.status_code, message, details)
        return response


def _drop_none(data: Dict[str, Any]) -> Dict[str, Any]:
    return {key: value for key, value in data.items() if value is not None}


def _parse_journal(entries: List[Dict[str, Any]]) -> List[JournalEntry]:
    return [
        JournalEntry(
            timestamp=entry['timestamp'],
            participant=entry.get('participant'),
            event=entry['event'],
            note=entry.get('note'),
            fingerprint=entry.get('fingerprint'),
            dedupe_count=entry['dedupe_count'],
            details=entry.get('details'),
        )
        for entry in entries
    ]


def _parse_issue_response(payload: Dict[str, Any]) -> IssueResponse:
    return IssueResponse(
        idempotency_key=payload['idempotency_key'],
        expires_at=payload['expires_at'],
        status=payload['status'],
    )


def _parse_verify_response(payload: Dict[str, Any]) -> VerifyResponse:
    return VerifyResponse(
        status=payload['status'],
        dedupe_count=payload['dedupe_count'],
        expires_at=payload['expires_at'],
        last_seen=payload['last_seen'],
        conflict_reason=payload.get('conflict_reason'),
        replay_hint=payload.get('replay_hint'),
        result=payload.get('result'),
        journal=_parse_journal(payload['journal']),
    )


def _parse_complete_response(payload: Dict[str, Any]) -> CompleteResponse:
    return CompleteResponse(
        status=payload['status'],
        result=payload.get('result'),
        journal=_parse_journal(payload['journal']),
    )


def _parse_record_view(payload: Dict[str, Any]) -> RecordView:
    return RecordView(
        status=payload['status'],
        tenant=payload['tenant'],
        action=payload['action'],
        resource=payload['resource'],
        fingerprint=payload.get('fingerprint'),
        dedupe_count=payload['dedupe_count'],
        created_at=payload['created_at'],
        last_seen=payload['last_seen'],
        expires_at=payload['expires_at'],
        conflict_reason=payload.get('conflict_reason'),
        result=payload.get('result'),
        journal=_parse_journal(payload['journal']),
    )

File: sdk/typescript/src/csiks.ts
export type RecordStatus = 'pending' | 'completed' | 'failed' | 'expired';
export type VerificationStatus =
  | 'accepted'
  | 'duplicate'
  | 'conflict'
  | 'expired'
  | 'completed'
  | 'failed';

export interface IssueRequest {
  tenant: string;
  action: string;
  resource: string;
  ttlSeconds?: number;
  issuedBy?: string;
  metadata?: unknown;
}

export interface IssueResponse {
  idempotencyKey: string;
  expiresAt: string;
  status: RecordStatus;
}

export interface VerifyRequest {
  tenant: string;
  action: string;
  resource: string;
  idempotencyKey: string;
  fingerprint?: string;
  participant?: string;
  dedupeToken?: string;
}

export interface VerifyResponse {
  status: VerificationStatus;
  dedupeCount: number;
  expiresAt: string;
  lastSeen: string;
  conflictReason?: string;
  replayHint?: string;
  result?: unknown;
  journal: JournalEntry[];
}

export interface CompleteRequest {
  tenant: string;
  action: string;
  resource: string;
  idempotencyKey: string;
  participant?: string;
  success: boolean;
  result?: unknown;
  note?: string;
}

export interface CompleteResponse {
  status: RecordStatus;
  result?: unknown;
  journal: JournalEntry[];
}

export interface JournalEntry {
  timestamp: string;
  participant?: string;
  event: JournalEvent;
  note?: string;
  fingerprint?: string;
  dedupeCount: number;
  details?: unknown;
}

export type JournalEvent =
  | 'issued'
  | 'verified'
  | 'completed'
  | 'failed'
  | 'conflict'
  | 'expired';

export interface RecordView {
  status: RecordStatus;
  tenant: string;
  action: string;
  resource: string;
  fingerprint?: string;
  dedupeCount: number;
  createdAt: string;
  lastSeen: string;
  expiresAt: string;
  conflictReason?: string;
  result?: unknown;
  journal: JournalEntry[];
}

export class CsiksError extends Error {
  public readonly status: number;
  public readonly details?: unknown;

  constructor(status: number, message: string, details?: unknown) {
    super(message);
    this.status = status;
    this.details = details;
  }
}

export interface CsiksClientOptions {
  baseUrl: string;
  headers?: Record<string, string>;
  fetchImpl?: typeof fetch;
}

const jsonHeaders = { 'content-type': 'application/json' } as const;

export class CsiksClient {
  private readonly baseUrl: string;
  private readonly headers: Record<string, string>;
  private readonly fetchImpl: typeof fetch;

  constructor(options: CsiksClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, '');
    this.headers = { ...options.headers };
    this.fetchImpl = options.fetchImpl ?? fetch;
  }

  async issueKey(request: IssueRequest): Promise<IssueResponse> {
    const response = await this.request('/keys/issue', {
      method: 'POST',
      body: JSON.stringify(toIssueWire(request)),
      headers: { ...jsonHeaders },
    });
    return fromIssueWire(await response.json());
  }

  async verifyKey(request: VerifyRequest): Promise<VerifyResponse> {
    const response = await this.request('/keys/verify', {
      method: 'POST',
      body: JSON.stringify(toVerifyWire(request)),
      headers: { ...jsonHeaders },
    });
    return fromVerifyWire(await response.json());
  }

  async completeKey(request: CompleteRequest): Promise<CompleteResponse> {
    const response = await this.request('/keys/complete', {
      method: 'POST',
      body: JSON.stringify(toCompleteWire(request)),
      headers: { ...jsonHeaders },
    });
    return fromCompleteWire(await response.json());
  }

  async getRecord(id: string): Promise<RecordView> {
    const response = await this.request(`/keys/${encodeURIComponent(id)}`, {
      method: 'GET',
    });
    return fromRecordWire(await response.json());
  }

  private async request(path: string, init: RequestInit): Promise<Response> {
    const response = await this.fetchImpl(`${this.baseUrl}${path}`, {
      ...init,
      headers: { ...this.headers, ...(init.headers as Record<string, string> | undefined) },
    });

    if (!response.ok) {
      let message = `Request failed with status ${response.status}`;
      let details: unknown;
      try {
        details = await response.clone().json();
        if (details && typeof details === 'object' && 'error' in (details as Record<string, unknown>)) {
          message = (details as Record<string, string>).error ?? message;
        }
      } catch {
        // swallow json errors - we keep default message
      }
      throw new CsiksError(response.status, message, details);
    }

    return response;
  }
}

interface IssueWireRequest {
  tenant: string;
  action: string;
  resource: string;
  ttl_seconds?: number;
  issued_by?: string;
  metadata?: unknown;
}

interface IssueWireResponse {
  idempotency_key: string;
  expires_at: string;
  status: RecordStatus;
}

interface VerifyWireRequest {
  tenant: string;
  action: string;
  resource: string;
  idempotency_key: string;
  fingerprint?: string;
  participant?: string;
  dedupe_token?: string;
}

interface VerifyWireResponse {
  status: VerificationStatus;
  dedupe_count: number;
  expires_at: string;
  last_seen: string;
  conflict_reason?: string;
  replay_hint?: string;
  result?: unknown;
  journal: JournalWireEntry[];
}

interface CompleteWireRequest {
  tenant: string;
  action: string;
  resource: string;
  idempotency_key: string;
  participant?: string;
  success: boolean;
  result?: unknown;
  note?: string;
}

interface CompleteWireResponse {
  status: RecordStatus;
  result?: unknown;
  journal: JournalWireEntry[];
}

interface RecordWireView {
  status: RecordStatus;
  tenant: string;
  action: string;
  resource: string;
  fingerprint?: string;
  dedupe_count: number;
  created_at: string;
  last_seen: string;
  expires_at: string;
  conflict_reason?: string;
  result?: unknown;
  journal: JournalWireEntry[];
}

interface JournalWireEntry {
  timestamp: string;
  participant?: string;
  event: JournalEvent;
  note?: string;
  fingerprint?: string;
  dedupe_count: number;
  details?: unknown;
}

const toIssueWire = (req: IssueRequest): IssueWireRequest => ({
  tenant: req.tenant,
  action: req.action,
  resource: req.resource,
  ttl_seconds: req.ttlSeconds,
  issued_by: req.issuedBy,
  metadata: req.metadata,
});

const toVerifyWire = (req: VerifyRequest): VerifyWireRequest => ({
  tenant: req.tenant,
  action: req.action,
  resource: req.resource,
  idempotency_key: req.idempotencyKey,
  fingerprint: req.fingerprint,
  participant: req.participant,
  dedupe_token: req.dedupeToken,
});

const toCompleteWire = (req: CompleteRequest): CompleteWireRequest => ({
  tenant: req.tenant,
  action: req.action,
  resource: req.resource,
  idempotency_key: req.idempotencyKey,
  participant: req.participant,
  success: req.success,
  result: req.result,
  note: req.note,
});

const fromIssueWire = (res: IssueWireResponse): IssueResponse => ({
  idempotencyKey: res.idempotency_key,
  expiresAt: res.expires_at,
  status: res.status,
});

const fromVerifyWire = (res: VerifyWireResponse): VerifyResponse => ({
  status: res.status,
  dedupeCount: res.dedupe_count,
  expiresAt: res.expires_at,
  lastSeen: res.last_seen,
  conflictReason: res.conflict_reason,
  replayHint: res.replay_hint,
  result: res.result,
  journal: res.journal.map(fromJournalWire),
});

const fromCompleteWire = (res: CompleteWireResponse): CompleteResponse => ({
  status: res.status,
  result: res.result,
  journal: res.journal.map(fromJournalWire),
});

const fromRecordWire = (res: RecordWireView): RecordView => ({
  status: res.status,
  tenant: res.tenant,
  action: res.action,
  resource: res.resource,
  fingerprint: res.fingerprint,
  dedupeCount: res.dedupe_count,
  createdAt: res.created_at,
  lastSeen: res.last_seen,
  expiresAt: res.expires_at,
  conflictReason: res.conflict_reason,
  result: res.result,
  journal: res.journal.map(fromJournalWire),
});

const fromJournalWire = (entry: JournalWireEntry): JournalEntry => ({
  timestamp: entry.timestamp,
  participant: entry.participant,
  event: entry.event,
  note: entry.note,
  fingerprint: entry.fingerprint,
  dedupeCount: entry.dedupe_count,
  details: entry.details,
});
